tests/test_parameters.py

```diff
@@ -1786,3 +1786,12 @@ def test_variable_api(self, simple_linear_model):
         new_values = np.random.rand(p.double_size)
         p.set_double_variables(new_values)
         np.testing.assert_allclose(p.get_double_variables(), new_values)
+
+class TestDiscountFactorParameter:
+    def test_discount_json(self):
+        """ Test loading a DiscountFactorParameter from JSON. """
+        model = load_model("discount.json")
+        # run model for period 2015-2020, with base year 2015 and discount rate of 0.035 (3.5%)
+        model.run()
+        # To add: test that the outputs of the parameter are the expected values.
```

Comment:
This needs to be done. You can look at the other tests for one that that uses the `@assert_rec` decorator. This allows you to write a simple function to return the value you expect the parameter to return.

Something like this:

```python
class TestDiscountFactorParameter:
    def test_discount_json(self):
        """ Test loading a DiscountFactorParameter from JSON. """
        model = load_model("discount.json")
        # run model for period 2015-2020, with base year 2015 and discount rate of 0.035 (3.5%)
        p = model.parameters['discount_factor']
        @assert_rec(model, p)
        def expected_func(timestep, scenario_index):
            year = timestep.year
            return 1/pow(1.035, year - 2015) 

        model.run()
```

Reply:
Thank you James, I have implemented this test now. How can I run the tests to make sure the test works as intended? I tried running `py.test tests` but I don't think that worked. I tried using wrong values so that it would fail the test but I did not get an error.

Reply:
`pytest tests` is the correct command.

Reply:
The test works for me. You can do `pytest tests/test_parameters.py -vv` if you want it to print the name of each test and whether it passed or failed. Or you can do `pytest tests -k discount` to filter for tests containing the word `discount`. 

And you can force a test failure with `assert False`

Reply:
Thanks James. So did you get a PASSED for the discount test? When I run `pytest tests/test_parameters.py -vv` I got FAILED for the `TestDiscountFactorParameter::test_discount_json`.

Here is the full message (also attached as screenshot). Since we have not cythonised the parameter yet, why do we get a KeyError in file `_parameters.pyx`?

```
>   @classmethod
E   KeyError: 'discountfactorparameter'

pywr/parameters/_parameters.pyx:1679: KeyError

During handling of the above exception, another exception occurred:

self = <test_parameters.TestDiscountFactorParameter object at 0x11f2837f0>

    def test_discount_json(self):
        """ Test loading a DiscountFactorParameter from JSON. """
>       model = load_model("discount.json")

tests/test_parameters.py:1794: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/helpers.py:17: in load_model
    model = Model.loads(data, path=path, solver=solver)
pywr/_model.pyx:189: in pywr._model.Model.loads
    ???
pywr/_model.pyx:386: in pywr._model.Model.load
    return components_data
pywr/_model.pyx:916: in pywr._model.listify.wrapper
    self.version = version
pywr/_model.pyx:382: in load_components
    def collect_components(data, key):
pywr/parameters/_parameters.pyx:1715: in pywr.parameters._parameters.load_parameter
    super(OffsetParameter, self).__init__(model, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   parameter = load_parameter(model, data.pop("parameter"))
E   TypeError: Unknown parameter type: "DiscountFactorParameter"

pywr/parameters/_parameters.pyx:1681: TypeError
```

<img width="941" alt="Screenshot 2020-06-24 at 11 42 02" src="https://user-images.githubusercontent.com/12497646/85532821-4dd8b080-b610-11ea-93d9-e63ea74f86ad.png">

Reply:
Yes, this traceback is confusing me also. I think you may need to re run `python setup.py develop` on Pywr. I think `_parameters.pyx` has changed since you last compiled it and the line numbers are wrong. The test passed on Travis.

Reply:
Thanks James, it has now passed the test on my machine as well.